[MIGRATION] Text references

hschink opened the pull request:
Implements [ticket 1991](https://www.openproject.org/issues/1991)

Commit:
Adds text reference migration

hschink commented:
This branch is ready for review because there is only one test failing that is not related to this branch.

ulferts commented:
I believe this should be refactored so that plugins can reuse the functionality. There might be plugins out there, e.g. the published meetings plugin (https://github.com/finnlabs/openproject-meeting) that might benefit from that.

Commit:
Extracts text reference update methods

hschink commented:
@ulferts: Is 13afb3d the kind of refactoring you had in mind?

ulferts commented on db/migrate/migration_utils/text_references.rb:
```diff
@@ -0,0 +1,161 @@
+      text
+    end
+
+    WORK_PACKAGE_LINK_REGEX = /(?<host>http:\/\/(\w|:)*)(\/timelines)?(\/projects\/(\w|-)*)?\/(issues|planning_elements)\/(?<id>\w*)/
```
This will have the following problems:

1) It will not find links with the https protocol.
2) It will change all links, regardless of whether they are internal or external. Only internal links should be replaced. An approach for that could be to require the Setting.host_name value to be included.
3) It will not find links to openproject instances that have a document root other than "/" defined (e.g. https://www.openproject.org/beta/issues/584)
4) It will not find links where the hostname includes a ".".

Commit:
Considers host name in settings

Commit:
Adds support for host postfixes

apfelfabrik commented on db/migrate/20130920053735_migrate_text_references_to_issues_and_planning_elements.rb:
```diff
@@ -0,0 +1,51 @@
+
+  COLUMNS_PER_TABLE = {
+    'boards' => ['description'],
+    'changesets' => ['comments'],
```
We're changing SVN commit messages here. Is that a good idea?

Commit:
Removes changeset migration

Changesets represent VCS commit messages. Thus, there is no point
migrating changesets because we do not migrate VCS commit messages as
well.

Commit:
Adds changelog entry

Commit:
Removes migration of text references in journals

Commit:
Makes test reference migration the last one

Commit:
Uses original timelines regex

Commit:
Resets journal primary key after migration

Commit:
Extends column update interface

Commit:
Ignores pre/code blocks

Commit:
Adds journal entry after text migration

apfelfabrik commented:
Just as a note, here's what I learned: The things one will most likely have to execute when migrating an old instance that had lots of former plugins:

```
rake migrations:timelines:reregister
rake migrations:strong_passwords:reregister
rake db:migrate # this will fail at some point, but before that schema version, subsequent rake tasks won't run.
rake migrations:attachments:delete_from_projects_and_versions # this is destructive
rake migrations:documents:delete # this, too, is destructive
rake db:migrate
```

If tasks somewhere in between fail, just `rake db:migrate` once more. With the schema a bit further ahead the other tasks might just work.

hschink commented:
@apfelfabrik I just rebased this PR against the current feature/rails3.
